pythia/models/butd.py

```diff
@@ -0,0 +1,294 @@
+            )
+            scores[:batch_size_t, t] = output
+
+            # Greedy decoding
+            if not teacher_forcing:
+                output_softmax = torch.log_softmax(output, dim=1)
+                _, indices = torch.max(output_softmax, dim=1, keepdim=True)
+                data["texts"] = torch.cat(
+                    (data["texts"], indices.view(batch_size_t, 1)), dim=1
+                )
+
+        targets = (
+            sample_list.answers[:, 0, 1:]
+            if not teacher_forcing
+            else sample_list.text[:, 1:]
+        )
+        sample_list.add_field("targets", targets)
+        model_output = {"scores": scores}
+        del (data)
+        gc.collect()
+
+        return model_output
+
+    def get_data_beam_search_t(self, data, batch_size_t):
+        data['image_features'] = data['image_features'][batch_size_t]
+        if 'state' in data:
+            h1 = data['state']['td_hidden'][0][batch_size_t]
+            c1 = data['state']['td_hidden'][1][batch_size_t]
+            h2 = data['state']['lm_hidden'][0][batch_size_t]
+            c2 = data['state']['lm_hidden'][1][batch_size_t]
+            data['state'] = {"td_hidden": (h1, c1), "lm_hidden": (h2, c2)}
+        return data
+
+    def beam_search(self, sample_list, k=5):
+        image_embedding_total  = self.process_feature_embedding("image", sample_list)
+
+        top_k_scores = torch.zeros(k, 1).to("cuda")
+        data = {}
+        data["texts"] = torch.LongTensor([[self.vocab.SOS_INDEX]] * k).to("cuda")
+        data["image_features"] = (
+            image_embedding_total.unsqueeze(1).expand(-1, k, -1, -1).squeeze(0)
+        )
+        seqs = data["texts"]
+        timesteps = self.text_processor.max_length
+
+        # Lists to store completed sequences and scores
+        complete_seqs = list()
+        complete_seqs_scores = list()
+
+        # Dummy output for loss calculation to not complain
+        outputs = torch.ones(sample_list.get_batch_size(), 52, self.vocab_size).to("cuda")
+
+        for t in range(timesteps):
+            data["texts"] = self.word_embedding(data["texts"].long())
+            if "state" not in data:
+                data["state"] = None
+
+            scores, data["state"] = self.decoder(
+                data["image_features"],
+                data["texts"][:, 0, :],
+                data["state"],
+            )
+
+            # Add predicted scores to top_k_scores
+            scores = torch.nn.functional.log_softmax(scores, dim=1)
+            scores = top_k_scores.expand_as(scores) + scores
+
+            # Find next top k scores and words
+            if t == 0:
+                top_k_scores, top_k_words = scores[0].topk(k, 0, True, True)
+            else:
+                top_k_scores, top_k_words = scores.view(-1).topk(
+                    k, 0, True, True
+                )
+
+            # Convert to vocab indices
+            prev_word_inds = top_k_words / self.vocab_size
```

Comment:
This will return a float, unless `top_k_words` is always divisible by `self.vocab_size`. Suggestion: `prev_word_inds = top_k_words // self.vocab_size`